parsers/IN.py

```diff
@@ -161,6 +219,125 @@ def fetch_consumption(
+            .ffill()
+        )
+        df_zone = df_npp.loc[df_npp["region"] == zone_key].copy()
+        df_zone = df_zone.loc[~df_zone.power_station.isna()]
+        df_zone = df_zone[df_zone.power_station.str.contains("TYPE:")]
+        df_zone = df_zone[["production_mode", "value"]]
+        df_zone = df_zone.groupby(["production_mode"]).sum()
+        production = {}
+        for mode in df_zone.index:
+            production[NPP_MODE_MAPPING[mode]] = round(
+                df_zone.iloc[df_zone.index.get_indexer_for([mode])[0]].get("value")
+                / CONVERSION_MWH_MW,
+                3,
+            )
+        return production
+    else:
+        raise ParserException(
+            parser="IN.py",
+            message=f"{target_datetime}: {zone_key} conventional production data is not available",
+        )
+
+
+def fetch_cea_production(
+    zone_key: str,
+    target_datetime: datetime,
+    session: Session = Session(),
+    logger: Logger = getLogger(__name__),
+) -> dict:
+    """Gets production data for wind, solar and other renewables
+    Other renewables includes a share of hydro, biomass and others and will categorized as unknown
+    DISCLAIMER: this data is only available since 2020/12/17"""
+    cea_link = "https://cea.nic.in/wp-content/uploads/daily_reports/{date:%d_%b_%Y}_Daily_Report.xlsx"
+    r: Response = session.get(cea_link.format(date=target_datetime))
+    if r.status_code == 200:
+        df_ren = pd.read_excel(r.url, engine="openpyxl", header=5)
+        df_ren = df_ren.rename(
+            columns={
+                df_ren.columns[1]: "region",
+                df_ren.columns[2]: "wind",
+                df_ren.columns[3]: "solar",
+                df_ren.columns[4]: "unknown",
+            }
+        )
+
+        df_ren.region = df_ren.region.str.strip()
+        df_ren.region = df_ren.region.map(CEA_REGION_MAPPING)
+        df_ren = df_ren.loc[df_ren.region == zone_key][["wind", "solar", "unknown"]]
+
+        dict_zone = df_ren.to_dict(orient="records")[0]
+        renewable_production = {
+            key: round(dict_zone[key] / CONVERSION_MWH_MW, 3) for key in dict_zone
+        }
+        return renewable_production
+    else:
+        raise ParserException(
+            parser="IN.py",
+            message=f"{target_datetime}: {zone_key} renewable production data is not available",
+        )
+
+
+def fetch_production(
+    zone_key: str,
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> dict:
+    if target_datetime is None:
+        target_datetime = arrow.now(tz=IN_NO_TZ).floor("day").datetime - timedelta(
+            days=2
+        )
+    else:
+        target_datetime = (
+            arrow.get(target_datetime).floor("day").datetime.replace(tzinfo=IN_NO_TZ)
+        )
+
+    renewable_production = fetch_cea_production(
+        zone_key=zone_key, session=session, target_datetime=target_datetime
```

Comment:
There are some type errors flagged here but I think they are safe to ignore based on the logic above. It complains about target_datetime potentially being None but that should not be possible since we replace it if it is.